src/tools/factory-prov-tool/WeaveFactoryProvisioningTool.py

```diff
@@ -0,0 +1,702 @@
+    return val if val is not None else defaultVal
+
+def _prefixLines(str, prefix):
+    return re.sub(r'.*', prefix + '\g<0>', str)
+
+def _hexDumpFile(file, prefix):
+    any(map(print, (prefix + (' '.join("0x%02X," % b for b in c)) for c in iter(lambda: file.read(16), b""))))
+    
+def _parseIntArg(val, min=None, max=None, argDesc='integer argument value', base=0):
+    try:
+        res = int(val, base=base)
+        if min != None and res < min:
+            res = None
+        elif max != None and res > max:
+            res = None
+    except ValueError:
+        res = None
+    if res == None:
+        raise _UsageError('Invalid %s: %s' % (argDesc, val))
+    return res
+
+def _readDataOrFileArg(argName, argValue, isBase64Funct=None):
+    if os.path.exists(argValue):
+        try:
+            with open(argValue, mode='rb') as f:
+                argValue = f.read()
+            if isBase64Funct != None and isBase64Funct(argValue):
+                argValue = base64.b64decode(argValue, validate=True)
+        except binascii.Error as ex:
+            raise _UsageError('Contents of %s file invalid: %s\nFile name: %s' % (argName, str(ex), deviceCert))
+        except IOError as ex:
+            raise _UsageError('Unable to read %s file: %s\nFile name: %s' % (argName, str(ex), deviceCert))
+    else:
+        try:
+            argValue = base64.b64decode(deviceCert, validate=True)
+        except binascii.Error as ex:
+            raise _UsageError('Invalid value specified for %s: %s' % (argName, str(ex)))
+    return argValue
+
+class _WipedNamedTemporaryFile(object):
+    def __init__(self, mode='w+b', suffix="", prefix=tempfile.template, dir=None):
+        if dir is None and os.path.isdir('/dev/shm'):
+            dir = '/dev/shm'
+        self.file = tempfile.NamedTemporaryFile(mode=mode, suffix=suffix, prefix=prefix, dir=dir, delete=True)
+        
+    def __getattr__(self, name):
+        file = self.__dict__['file']
+        return getattr(file, name)
+
+    def __enter__(self):
+        self.file.__enter__()
+        return self
+
+    def close(self):
+        self._wipe()
+        self.file.close()
+
+    def __del__(self):
+        self._wipe()
+        self.file.__del__()
+
+    def __exit__(self, exc, value, tb):
+        self._wipe()
+        return self.file.__exit__(exc, value, tb)
+    
+    def _wipe(self):
+        if os.path.isfile(self.file.name):
+            fileLen = os.stat(self.file.name).st_size
+            with open(self.file.name, mode='wb') as f:
+                lenWiped = 0
+                while lenWiped < fileLen:
+                    f.write(b'\0' * 16)
+                    lenWiped += 16
+                f.truncate()
+
+def isBase64WeaveCert(s):
+    return s.startswith(b"1QAABAAB")
+
+def isBase64WeavePrivateKey(s):
+    return s.startswith(b"1QAABAAC") or s.startswith(b"1QAABAAD")
+
+def encodeESP32FirmwareImage(segments, entryPoint=0, flashMode=0, flashFreq=0, flashSize=0):
```

Comment:
There seem to be ESP32, J-Link, and nRFisms in this script. Is there a plan or vision to scale this or to support a plug-in style model where the platform-specific bits of this live in device layer?

Otherwise, this doesn't seem like it scales well to other platforms.

Reply:
Extending this module to support other SoCs is straightforward, in some cases even trivial.

ESP32 is of course its own thing, unlikely to share an implementation with any other platform.  However, if a new platform is supported by SEGGER J-Link, adding it is as simple as defining a new instance of the Target class (similar to Target_nrf52840) and adding it to the Targets dictionary.  Beyond the existence of the Target_nrf52840 object, there's nothing nRF5-specific about this code.

I also believe it will be straightforward to add a controller class for OpenOCD.  This would allow JTAG and SWD based data-injection for any supported platform without the need for J-Link hardware.

Reply:
@gerickson : Did my response adequately address your question?

Reply:
Yes, I believe so.